<a href="https://colab.research.google.com/github/VivianeMatosOliveira/Curso_Analista_Dados/blob/main/Exercicios/modulo_22_exercicio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/main/media/logo/newebac_logo_black_half.png" alt="ebac-logo">

---

# **Módulo** | Análise de Dados: Fundamentos de Estatística
Caderno de **Exercícios**<br> 
Professor [André Perez](https://www.linkedin.com/in/andremarcosperez/)

---

# **Tópicos**

<ol type="1">
  <li>Média e Variância;</li>
  <li>Ordem e Posição;</li>
  <li>Correlação.</li>
</ol>

---

# **Exercícios**

## 1\. Tráfego de São Paulo

Neste exercício, vamos continuar a analisar os dados de mobilidade urbana da cidade de São Paulo. A base de dados está neste [link](https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/develop/dataset/traffic.csv) e é uma copia do dado original, presente neste [link](https://archive.ics.uci.edu/ml/datasets/Behavior+of+the+urban+traffic+of+the+city+of+Sao+Paulo+in+Brazil). A base de dados contem a quantidade de acidentes ocorridos na cidade entre 14/12/09 e 18/12/09, das 07:00h ás 20:00h, agregados em intervalos de 30 minutos.

In [ ]:
!wget -q "https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/develop/dataset/traffic.csv" -O traffic.csv

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('traffic.csv', sep=';')

In [ ]:
df.head()

hour  bus_stoped  broken_truck  vehicle_excess  accident_victim  \
0     1           0             0               0                0   
1     2           0             0               0                0   
2     3           0             0               0                0   
3     4           0             0               0                0   
4     5           0             0               0                0   

   running_over  fire_vehicles  occurrence_freight  \
0             0              0                   0   
1             0              0                   0   
2             0              0                   0   
3             0              0                   0   
4             0              0                   0   

   incident__dangerous_freight  lack_electricity  fire  flooding  \
0                            0                 0     0         0   
1                            0                 0     0         0   
2                            0                 0     0         0   
3                            0                 0     0         0   
4                            0                 0     0         0   

   manifestations  defect_trolleybuses  tree_road  semaphore_off  \
0               0                    0          0              0   
1               0                    0          0              0   
2               0                    0          0              0   
3               0                    0          0              0   
4               0                    0          0              0   

   intermittent_emaphore slowness_traffic_%  
0                      0                4,1  
1                      0                6,6  
2                      0                8,7  
3                      0                9,2  
4                      0               11,1

- **Exemplo**: (para relembrar)

O código abaixo extrai a 13ª linha do arquivo que representa a 14ª meia hora do dia 14/12/09, contadas a partir das 07:00h, ou seja, todos os incidentes ocorridos na cidade no dia em questão entre as 13:30h e 14:00h. Vemos, por exemplo, que dois ônibus e dois caminhões ficaram paradados nas ruas da cidade, atrapalhando o tráfego.

In [ ]:
df.iloc[[13]]

hour  bus_stoped  broken_truck  vehicle_excess  accident_victim  \
13    14           2             2               0                1   

    running_over  fire_vehicles  occurrence_freight  \
13             0              0                   0   

    incident__dangerous_freight  lack_electricity  fire  flooding  \
13                            0                 0     0         0   

    manifestations  defect_trolleybuses  tree_road  semaphore_off  \
13               0                    0          0              0   

    intermittent_emaphore slowness_traffic_%  
13                      0                4,9

### **1.1. Agregação** 

Neste primeira atividade, você deve gerar um array NumPy por dia. Para cada array você deve somar todos os incidentes que aconteceram naquela meia hora. Sendo assim, cada array deve ter 27 posições, cada qual com a soma dos incidentes daquela meira hora.

**Dica**: Você deve remover a primeira e a última coluna.

**Dica**: Os arrays devem ter os seguintes valores:

```python
14: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 5, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 2, 3]
15: [3, 0, 0, 0, 2, 4, 2, 1, 3, 0, 2, 0, 2, 3, 2, 2, 4, 7, 6, 3, 0, 1, 1, 5, 4, 6, 0]
16: [1, 1, 2, 4, 4, 5, 3, 2, 4, 2, 1, 1, 3, 6, 2, 0, 2, 8, 4, 6, 4, 1, 2, 7, 15, 4, 0]
17: [4, 1, 1, 0, 0, 0, 2, 3, 2, 7, 2, 1, 7, 3, 6, 4, 3, 2, 2, 3, 2, 5, 2, 0, 1, 4, 15]
18: [10, 1, 1, 5, 5, 7, 3, 4, 3, 1, 2, 3, 2, 2, 6, 2, 2, 4, 2, 3, 0, 3, 3, 2, 3, 6, 0]
```

In [ ]:
# resposta da questão 1.1

In [ ]:
import numpy as np


In [ ]:
traffic_array = np.array(df)
day = 14
timebox = 0
incidents = []
incident_by_day = dict()

#incident_array = np.array(df.iloc[:, 0:17])
#incidents = incidents + incident_array.astype(np.int)

for item in traffic_array:
  incidents_array = item[1:len(item) -1].astype(int).sum()
  incidents.append(incidents_array)
  timebox = timebox + 1
  incident_by_day [day] = incidents

  try:
    if len (incidents) == 27:
      day = day + 1
      incidents = []

  except ValueError:
    continue


for day in incident_by_day:
  print(f'{day}: {incident_by_day[day]}')



14: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 5, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 2, 3]
15: [3, 0, 0, 0, 2, 4, 2, 1, 3, 0, 2, 0, 2, 3, 2, 2, 4, 7, 6, 3, 0, 1, 1, 5, 4, 6, 0]
16: [1, 1, 2, 4, 4, 5, 3, 2, 4, 2, 1, 1, 3, 6, 2, 0, 2, 8, 4, 6, 4, 1, 2, 7, 15, 4, 0]
17: [4, 1, 1, 0, 0, 0, 2, 3, 2, 7, 2, 1, 7, 3, 6, 4, 3, 2, 2, 3, 2, 5, 2, 0, 1, 4, 15]
18: [10, 1, 1, 5, 5, 7, 3, 4, 3, 1, 2, 3, 2, 2, 6, 2, 2, 4, 2, 3, 0, 3, 3, 2, 3, 6, 0]


### **1.2. Métricas** 

Para cada array você deve calcular as seguintes métricas: 

 - **média**;
 - **desvio padrão**.

In [ ]:
# resposta da questão 1.2

In [ ]:
type(incident_by_day)

dict

In [ ]:
lista = []  # calculando a média do período
      
for value in incident_by_day.values():
    lista.append(value) 
          
media = np.mean(lista) 
      
print(media) 

2.5185185185185186


In [ ]:
list2 = [] # Calculando o desvio padrão do período
      
for value in incident_by_day.values(): 
    list2.append(value) 
          
std = np.std(list2).round(2)
      
print(std) 

2.62


In [ ]:
dic_df = pd.DataFrame(incident_by_day) # criando um dataFrame a partir dos dados contidos no dicionário - incident_by_day
dic_df.head()

14  15  16  17  18
0   0   3   1   4  10
1   0   0   1   1   1
2   0   0   2   1   1
3   0   0   4   0   5
4   0   2   4   0   5

### **Calculando a média por dia**

In [ ]:
mean_day_14 = dic_df[14].mean().round(2)
print(mean_day_14)

0.59


In [ ]:
mean_day_15 = dic_df[15].mean().round(2)
print(mean_day_15)

2.33


In [ ]:
mean_day_16 = dic_df[16].mean().round(2)
print(mean_day_16)

3.48


In [ ]:
mean_day_17 = dic_df[17].mean().round(2)
print(mean_day_17)

3.04


In [ ]:
mean_day_18 = dic_df[18].mean().round(2)
print(mean_day_18)

3.15


### **Calculando o desvio padrão por dia**

In [ ]:
std_day_14 = dic_df[14].std().round(2)
print(std_day_14)

1.22


In [ ]:
std_day_15 = dic_df[15].std().round(2)
print(std_day_15)

2.06


In [ ]:
std_day_16 = dic_df[16].std().round(2)
print(std_day_16)

3.11


In [ ]:
std_day_17 = dic_df[17].std().round(2)
print(std_day_17)

3.11


In [ ]:
std_day_18 = dic_df[18].std().round(2)
print(std_day_18)

2.23


### **1.3. Interpretação** 

Baseado nos resultados da questão 1.2, responda:

> Qual dia apresenta a maior média de acidentes por meia hora?

**Resposta:** dia 16 com média de 3.48 acidentes

> Qual dia apresenta a menor variação de acidentes por meia hora?

**Resposta:** dia 14 com variação de 1.22 

---